src/v/cluster/members_backend.cc

```diff
@@ -170,17 +187,34 @@ bool is_in_replica_set(
 }
 
 ss::future<> members_backend::calculate_reallocations(update_meta& meta) {
-    switch (meta.update.type) {
+    // on demand update
+    if (!meta.update) {
+        co_await calculate_reallocations_for_even_partition_count(
+          meta, partition_allocation_domains::consumer_offsets);
+        if (
+          meta.partition_reallocations.size()
+          < _max_concurrent_reallocations()) {
+            co_await calculate_reallocations_for_even_partition_count(
+              meta, partition_allocation_domains::common);
+        }
+    }
+    // update caused by node event i.e. addition/decommissioning/recommissioning
+    switch (meta.update->type) {
```

Comment:
given that `meta.update` could be nullopt (check right above this) it doesn't seem like anything guarantees that after calculate_reallocations_for_even_partition_count it has a value here when the optional is dereferenced?

Reply:
good point, i missed that